Build fail with 5.4.0

updated from 5.3.1 to 5.4.0, then gradle build fails. 'Clean Project' returns the same exception.

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':myapp'.
> Cannot change dependencies of configuration ':myapp:api' after it has been included in dependency resolution.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s
```

My project gradle
```
buildscript {
    ext.kotlin_version = '1.2.60-eap-27'
    repositories {
        google()
        jcenter()
        maven { url 'https://dl.bintray.com/kotlin/kotlin-eap' }
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.2.0-beta04'
        classpath 'io.realm:realm-gradle-plugin:5.4.0'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath 'com.google.gms:google-services:4.0.1'
    }
}

allprojects {

    repositories {
        google()
        jcenter()
        maven { url 'http://clojars.org/repo' }
        maven { url 'https://dl.bintray.com/kotlin/kotlin-eap' }
    }
}

ext {
    minSdk = 17
    targetSdk = 28
    compileSdk = 28
    supportLib = '27.1.1'
    fb = '4.34.0'
    glide = '4.7.1'
}

task clean(type: Delete) {
    delete rootProject.buildDir
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,10 @@
+## 5.4.1 (YYYY-MM-DD)
+
+### Bug Fixes
+
+* Fix Realm Gradle Plugin adding depencies in a way incompatible with Kotlin Android Extensions. This was introduced in Realm Java 5.4.0 (#6080).
+
+
 ## 5.4.0 (2018-07-22)
 
 ### Enhancements

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,7 +2,7 @@
 
 ### Bug Fixes
 
-* Fix Realm Gradle Plugin adding depencies in a way incompatible with Kotlin Android Extensions. This was introduced in Realm Java 5.4.0 (#6080).
+* Fix Realm Gradle Plugin adding dependencies in a way incompatible with Kotlin Android Extensions. This was introduced in Realm Java 5.4.0 (#6080).
 
 
 ## 5.4.0 (2018-07-22)

--- a/gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy
+++ b/gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy
@@ -127,6 +127,9 @@ class Realm implements Plugin<Project> {
         return !hasAnnotationProcessorConfiguration
     }
 
+    // This will setup the required dependencies.
+    // Due to how Gradle works, we have no choice but to run this code every time any of the parameters
+    // in the Realm extension is changed.
     private static void setDependencies(Project project, String dependencyConfigurationName, boolean syncEnabled, boolean kotlinExtensionsEnabled) {
         // remove libraries first
         def iterator = project.getConfigurations().getByName(dependencyConfigurationName).getDependencies().iterator()

--- a/gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy
+++ b/gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy
@@ -19,6 +19,7 @@ package io.realm.gradle
 import com.android.build.gradle.AppPlugin
 import com.android.build.gradle.LibraryPlugin
 import com.neenbedankt.gradle.androidapt.AndroidAptPlugin
+import io.realm.gradle.RealmPluginExtension
 import io.realm.transformer.RealmTransformer
 import org.gradle.api.GradleException
 import org.gradle.api.Plugin
@@ -47,8 +48,20 @@ class Realm implements Plugin<Project> {
         def hasAnnotationProcessorConfiguration = project.getConfigurations().findByName('annotationProcessor') != null
         // TODO add a parameter in 'realm' block if this should be specified by users
         def preferAptOnKotlinProject = false
-
+        def dependencyConfigurationName = getDependencyConfigurationName(project)
         def extension = project.extensions.create('realm', RealmPluginExtension)
+        extension.addPropertyListener(RealmPluginExtension.KEY_KOTLIN_EXTENSIONS_ENABLED, new RealmPluginExtension.PropertyChangedListener<Boolean>() {
+            @Override
+            void onChange(Boolean checked) {
+                setDependencies(project, dependencyConfigurationName, extension.syncEnabled, extension.kotlinExtensionsEnabled)
+            }
+        })
+        extension.addPropertyListener(RealmPluginExtension.KEY_SYNC_ENABLED, new RealmPluginExtension.PropertyChangedListener<Boolean>() {
+            @Override
+            void onChange(Boolean checked) {
+                setDependencies(project, dependencyConfigurationName, extension.syncEnabled, extension.kotlinExtensionsEnabled)
+            }
+        })
         extension.kotlinExtensionsEnabled = useKotlinExtensionsDefault
 
         if (shouldApplyAndroidAptPlugin(usesAptPlugin, isKotlinProject,
@@ -58,7 +71,6 @@ class Realm implements Plugin<Project> {
         }
 
         project.android.registerTransform(new RealmTransformer(project))
-        def dependencyConfigurationName = getDependencyConfigurationName(project)
 
         project.repositories.add(project.getRepositories().jcenter())
         project.dependencies.add(dependencyConfigurationName, "io.realm:realm-annotations:${Version.VERSION}")
@@ -73,8 +85,6 @@ class Realm implements Plugin<Project> {
             project.dependencies.add("annotationProcessor", "io.realm:realm-annotations-processor:${Version.VERSION}")
             project.dependencies.add("androidTestAnnotationProcessor", "io.realm:realm-annotations-processor:${Version.VERSION}")
         }
-
-        setDependencies(project, dependencyConfigurationName, extension.syncEnabled, extension.kotlinExtensionsEnabled)
     }
 
     private static boolean isTransformAvailable() {

--- a/gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy
+++ b/gradle-plugin/src/main/groovy/io/realm/gradle/Realm.groovy
@@ -74,9 +74,7 @@ class Realm implements Plugin<Project> {
             project.dependencies.add("androidTestAnnotationProcessor", "io.realm:realm-annotations-processor:${Version.VERSION}")
         }
 
-        project.afterEvaluate {
-            setDependencies(project, dependencyConfigurationName, extension.syncEnabled, extension.kotlinExtensionsEnabled)
-        }
+        setDependencies(project, dependencyConfigurationName, extension.syncEnabled, extension.kotlinExtensionsEnabled)
     }
 
     private static boolean isTransformAvailable() {
@@ -121,7 +119,6 @@ class Realm implements Plugin<Project> {
 
     private static void setDependencies(Project project, String dependencyConfigurationName, boolean syncEnabled, boolean kotlinExtensionsEnabled) {
         // remove libraries first
-        
         def iterator = project.getConfigurations().getByName(dependencyConfigurationName).getDependencies().iterator()
         while (iterator.hasNext()) {
             def item = iterator.next()

--- a/realm-transformer/src/main/java/io/realm/gradle/RealmPluginExtension.java
+++ b/realm-transformer/src/main/java/io/realm/gradle/RealmPluginExtension.java
@@ -18,8 +18,54 @@ package io.realm.gradle;
 
 import org.gradle.api.tasks.Input;
 
+import java.util.LinkedHashMap;
+import java.util.Map;
+
 public class RealmPluginExtension {
 
-    @Input public boolean syncEnabled = false;
-    @Input public boolean kotlinExtensionsEnabled = false;
+    public static final String KEY_SYNC_ENABLED = "syncEnabled";
+    public static final String KEY_KOTLIN_EXTENSIONS_ENABLED = "kotlinExtensionsEnabled";
+
+    private boolean syncEnabled;
+    private boolean kotlinExtensionsEnabled;
+    private Map<String, PropertyChangedListener> listeners = new LinkedHashMap<>();
+
+    @Input
+    public boolean isSyncEnabled() {
+        return syncEnabled;
+    }
+
+    public void setSyncEnabled(boolean syncEnabled) {
+        this.syncEnabled = syncEnabled;
+        notifyChange(KEY_SYNC_ENABLED, syncEnabled);
+    }
+
+    @Input
+    public boolean isKotlinExtensionsEnabled() {
+        return kotlinExtensionsEnabled;
+    }
+
+    public void setKotlinExtensionsEnabled(boolean kotlinExtensionsEnabled) {
+        this.kotlinExtensionsEnabled = kotlinExtensionsEnabled;
+        notifyChange(KEY_KOTLIN_EXTENSIONS_ENABLED, kotlinExtensionsEnabled);
+    }
+
+    public void addPropertyListener(String property, PropertyChangedListener listener) {
+        listeners.put(property, listener);
+    }
+
+    private void notifyChange(String key, Object value) {
+        PropertyChangedListener listener = listeners.get(key);
+        if (listener != null) {
+            // Up to users of the API to use the correct generic type, otherwise it will crash
+            // at runtime.
+            //noinspection unchecked
+            listener.onChange(value);
+        }
+    }
+
+    // Callback triggered when the extension property is changed
+    public interface PropertyChangedListener<T> {
+        void onChange(T value);
+    }
 }

--- a/realm-transformer/src/main/java/io/realm/transformer/Utils.java
+++ b/realm-transformer/src/main/java/io/realm/transformer/Utils.java
@@ -73,7 +73,7 @@ public class Utils {
 
     public static boolean isSyncEnabled(Project project) {
         RealmPluginExtension realmExtension = (RealmPluginExtension) project.getExtensions().findByName("realm");
-        return realmExtension != null && realmExtension.syncEnabled;
+        return realmExtension != null && realmExtension.isSyncEnabled();
     }
 
     public static List<File> getBootClasspath(Project project) {

